src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4707,8 +4708,14 @@ void StorageReplicatedMergeTree::read(
 
         auto cluster = local_context->getCluster(local_context->getSettingsRef().cluster_for_parallel_replicas);
 
-        Block header =
-            InterpreterSelectQuery(modified_query_ast, local_context, SelectQueryOptions(processed_stage).analyze()).getSampleBlock();
+        Block header;
+
+        if (local_context->getSettingsRef().allow_experimental_analyzer)
+            header = InterpreterSelectQueryAnalyzer::getSampleBlock(
```

Comment:
This is completely incorrect. It uses a query for shards created by the old infrastructure. We need to do here almost the same as in `StorageDistributed::read`, but it doesn't make sense before we merge #50097.

Reply:
I'm missing most of the context, so sorry if this is a stupid question. Do you mind elaborating about the differences about the old and the new infrastructure? Would this be something to be implemented in ClusterProxy::rewriteSelectQuery to make it similar to `buildQueryTreeDistributed` but for the pure parallel replicas case?

Reply:
`ClusterProxy::rewriteSelectQuery` works on top of AST. The intention of the new analyzer is to avoid modifying AST at all. So yes, basically we need to split logic inside of `buildQueryTreeDistributed` into parts and reuse it here.

Reply:
Makes absolute sense. Although I guess that change is additional to the change proposed here, which is just changing the way the header is calculated depending on the setting in a similar way it's done in other places.